### PR TITLE
Better error handling on the backend side

### PR DESCRIPTION
**Pull request by artem-smotrakov**
Currently, the backend returns "internal server error" even if there was not actually a problem on the server side. This PR tries to improve error handling in KeyConjurer so that a user gets more info in errors returned by KeyConjurere.

Main updates:
* Introduced standard errors for the Okta client. The client now inspects errors from Okta a bit deeper and wraps them into one of the standard errors.
* Introduced standard errors for authentication providers. The Okta provider now also inspects Okta's errors a bit deeper and returns one of the standard errors.
* When getting SAML assertion, the lambda handler now checks errors from the authentication provider to pick up a right error code.
* Added several tests.

### CODE REVIEW

File: api/authenticators/okta/okta_auth.go [hunk 1]
```diff
@@ -83,14 +83,14 @@ func (o *oktaAuthClient) do(ctx context.Context, method, path string, data, resu
 
 	b, err := json.Marshal(data)
 	if err != nil {
-		return
+		return nil, WrapError(err, ErrOktaBadRequest)
```
Comment (trinitroglycerin) [inline api/authenticators/okta/okta_auth.go:86]: Because this function has a return signature with aliases in:

````go
func (*oktaAuthClient) do(...) (response *http.Response, err error)
```

We don't need to explicitly return two values every time. The following would suffice:

````go
if err != nil {
    err = WrapError(err, ErrOktaBadRequest)
    return
}
````

File: api/authenticators/okta/okta_auth.go [hunk 2]
```diff
@@ -305,20 +304,82 @@ func (o *oktaAuthClient) GetSAMLResponse(ctx context.Context, application okta.A
...
+	ErrOktaUnauthorized          OktaError = errors.New("unauthorized")
+	ErrOktaForbidden             OktaError = errors.New("forbidden")
+	ErrOktaCouldNotCreateSession OktaError = errors.New("could not create a session")
+	ErrOktaSAMLError             OktaError = errors.New("could not get a SAML response")
+	ErrOktaInternalServerError   OktaError = errors.New("internal server error")
+	ErrOktaUnspecified           OktaError = errors.New("unspecified")
+)
+
+// WrapError wraps an error into a standard Okta client error.
+func WrapError(err error, oktaErr OktaError) error {
```
Comment (trinitroglycerin) [inline api/authenticators/okta/okta_auth.go:345]: I don't think this needs to be exported. You should rename this to `wrapError` so it is not visible outside of this package.

File: api/authenticators/okta/okta_auth.go [hunk 3]
```diff
@@ -305,20 +304,82 @@ func (o *oktaAuthClient) GetSAMLResponse(ctx context.Context, application okta.A
...
+	return fmt.Errorf("%w: %s", oktaErr, err.Error())
+}
+
+// GetMessage extracts a message from an error.
+// If it is an error from Okta, the function returns its summary.
+// Otherwise, it returns err.Error().
+func GetMessage(err error) string {
```
Comment (trinitroglycerin) [inline api/authenticators/okta/okta_auth.go:352]: As above, this should not be exported.

**Comment by artem-smotrakov** [comment]
Thanks for the review @trinitroglycerin ! The comments make sense to me. In fact, there are a few more functions that should not be exported. I've addressed your comments.

File: api/authenticators/okta/authenticator.go [hunk 4]
```diff
@@ -95,19 +95,19 @@ func extractRole(group *okta.Group) (extractedRole, bool) {
 // This will initiate a multi-factor request with Duo.
 func (a *Authenticator) GenerateSAMLAssertion(ctx context.Context, creds core.Credentials, appID string) (*core.SAMLResponse, error) {
 	if appID == "" {
-		return nil, errors.New("appID cannot be an empty string")
+		return nil, core.NewBadRequestError("appID cannot be an empty string")
```
Comment (trinitroglycerin) [inline api/authenticators/okta/authenticator.go:98]: I know we discussed this in our 1-1 to try to prevent having multiple translation layers.

Instead of having every Authenticator provide a specific set of errors I would not recommend having the translation layer 'here'.

The return value of any given type, and what kind of things it can return, are defined by its type signature. Even if we are ensuring that this particular implementation always returns a specific set of errors by using these `core.NewXError` functions, other implementations might not, and any implementation is free to provide any kind of error because the type signature allows for it.

Constructor patterns for values in general should also be avoided, because it implies that a value cannot be created or useful without a function (which goes against Go's policy of making the zero value useful), and I can't say I've seen many examples where a package would create constructor functions for errors so that they can be used by other packages.

My recommendation here would be to return either Sentinel values, or struct values, and have the caller of this function translate them there, rather than polluting the authenticator logic with the 'core' package; there _are_ other examples of this package depending on the core package, but that is part of the reason why KeyConjurer is so hard to refactor, as everything is very tightly coupled.

TL;DR - I would recommend changing this and other errors in here to be something like this:


````go
type OktaErrorReason struct{}

var (
  ReasonAppIDEmptyString = OktaErrorReason{}
  ...
)

type OktaAuthenticatorError struct {
  Reason OktaErrorReason
  Inner error
}

func (o OktaAuthenticatorError) Error() string {
  switch o.Reason {
  case ReasonAppIDEmptyString:
    return "appID cannot be an empty string"
  /// others here
  }
}

func (o OktaAuthenticatorError) Unwrap() error {
  return o.Inner
}

func GenerateSAMLAssertion(..) (...) {
  if appID == "" {
    return nil, OktaAuthenticatorError{Reason: ReasonAppIDEmptyString}
  }

  st, err := a.oktaAuthClient.Authn(...)
  if err != nil {
    return nil, OktaAuthenticatorError{Reason: translateErrorToReason(err), Inner: err}
  }

  ....
}

func translateErrorToReason(err error) OktaErrorReason {
   if errors.Is(err, OktaBadRequest) {
     return ReasonBadRequest
   }
}
````

We can then, later on, enumerate all of the known provider errors in `GetErrorCode()`, as is currently done. I recognise that that will increase the size of `GetErrorCode()`, but IMO having the one keyconjurer specific translation layer right before we send responses back to the user is more desirable than having values being altered as they pass through the stack multiple times to conform to an interface specifically to deliver response codes.

Comment (artem-smotrakov) [reply api/authenticators/okta/authenticator.go:98]: Thanks for the feedback @trinitroglycerin !

> The return value of any given type, and what kind of things it can return, are defined by its type signature. Even if we are ensuring that this particular implementation always returns a specific set of errors by using these `core.NewXError` functions, other implementations might not, and any implementation is free to provide any kind of error because the type signature allows for it.

I actually meant to introduce a contract for errors returned by authentication providers. A provider is free to return any error, but if it wants others to understand it, it should wrap its errors into one of the standard errors.

> Constructor patterns for values in general should also be avoided, because it implies that a value cannot be created or useful without a function (which goes against Go's policy of making the zero value useful), and I can't say I've seen many examples where a package would create constructor functions for errors so that they can be used by other packages.

`core.NewXError` methods can be removed, it is no problem. The provider can just wrap errors/messages itself.

> My recommendation here would be to return either Sentinel values, or struct values, and have the caller of this function translate them there, rather than polluting the authenticator logic with the 'core' package; there _are_ other examples of this package depending on the core package, but that is part of the reason why KeyConjurer is so hard to refactor, as everything is very tightly coupled.

> We can then, later on, enumerate all of the known provider errors in `GetErrorCode()`, as is currently done. I recognise that that will increase the size of `GetErrorCode()`, but IMO having the one keyconjurer specific translation layer right before we send responses back to the user is more desirable than having values being altered as they pass through the stack multiple times to conform to an interface specifically to deliver response codes.

If I understand correctly, the dependency to the `core` package exists naturally because `AuthenticationProvider` is defined in `core`. I think `core` is a right place for defining errors for providers. The fact that we have the `AuthenticationProvider` interface made me choose the way with standard errors on the provider layer. If we didn't have this interface, I'd probably go with analysing errors in the lambda handler. Another reason is splitting and spreading responsibilities for analysing errors. The Okta provider knows it talks to Okta and it should be aware of its errors. It should have all the info to select a right error according to the contract. The lambda knows that authentication providers should return a standard error, so its only task is to be able to translate a standard error to an error code.

I'd stay with the current layered approach because:
- It would avoid lambda hander's "dependency" on the implementation of specific authentication providers like Okta. Otherwise, a consolidated `GetErrorCode()` would need to take into account Okta's and other errors.
- Error handling can be unit-tested separately for each layer.

I admit that the downside is that the logic is more spread between the classes which may be a bit more difficult to follow without keeping in mind the contract for errors on each layer.

I also admit that I have little experience in Go, so I can easily be missing something important. If you think, this is absolutely necessary to change the design, I'll do that. Maybe we can also have a third opinion.

Comment (trinitroglycerin) [reply api/authenticators/okta/authenticator.go:98]: > If I understand correctly, the dependency to the core package exists naturally because AuthenticationProvider is defined in core

The providers have a dependency on core so they can do a compile-time check in the provider bundle itself to ensure that they correctly satisfy the Authenticator interface:

````go
var _ core.AuthenticationProvider = &Authenticator{}
````

And also so that they have access to the shared type `SAMLResponse`.

In hindsight, I think this was a mistake, and a package named 'core' is too broadly scoped such that it will end up with a lot of really strange spaghetti issues. That's not your fault, though, and I don't think that should prevent this PR from being merged nor should - despite my previous comment - the burden be on you to alter that in this PR.

> I admit that the downside is that the logic is more spread between the classes which may be a bit more difficult to follow without keeping in mind the contract for errors on each layer.

I feel like if we are going to have contractual obligations where each error returned from the authentication provider must be of a specific type, then we should enforce that with the type system; instead of returning `error`, we should return `AuthenticationProviderError`, or some other kind of struct. Otherwise, we should accept that authentication providers can return any particular error, and adjust our error handling code accordingly.

I think this might be a good middle-ground, although it feels not very Go-like. I'll have to study some other examples of Go code where a similar situation like this is used. I agree that handling Okta specific errors at the tip of the application _in general_ is not ideal.

One could also make the argument that the structure of this project leads it to be more generic than it needs to be. We have the authentication provider interface, but _realistically_ we only support Okta and will never support two authentication providers at the same time, and there's a lot of complexity added to the project by trying to support that. We _did_ support multiple providers at the same time during the transition from Okta to OneLogin, but the authentication provider interface didn't really help much with this, and it would have been better to create an entirely new flow for Okta rather than having it be supported by the same flows as OneLogin in the past, though that's an aside.

I feel for the time being, the state the PR is in is just fine for it to be merged. I think my concerns come out of other issues to do with the complexity of this project compared to what it achieves, not your particular changes.

Comment (artem-smotrakov) [reply api/authenticators/okta/authenticator.go]: > I feel like if we are going to have contractual obligations where each error returned from the authentication provider must be of a specific type, then we should enforce that with the type system; instead of returning error, we should return AuthenticationProviderError, or some other kind of struct. Otherwise, we should accept that authentication providers can return any particular error, and adjust our error handling code accordingly.

Makes sense to me. I've updated the `AuthenticationProvider` interface. I've also removed two `core.NewXError` functions and renamed the third one. I'm going to merge it later today if no objections.
